Numerous adjustments to Deardorff model, change sign convention for tau

ewquon opened the pull request:
Deardorff:
* fix sign error on buoyancy term
* add limit on mixing length under stable stratification -- should be less than or equal to the cube root of cell volume used for unstable conditions and, following WRF, set a lower bound of 1e-8
* add limit on `rho*e` (> eps) after slow step
* add "wall effect" to the dissipation coefficient, enable by setting `erf.Ce_wall = 3.9`
* initialize SFS TKE field (`e` rather than `rho*e`) to a constant by setting `erf.KE_0`, default is to 0.1 (match SOWFA / AMR-Wind) -- for backwards compatibility, can still set `erf.RhoKE_0`
* update eddy diffusivities in first RK stage -- _work in progress..._

General turbulence modeling consistency:
* flip sign on tau so that `tau_ij = R_ij - R_kk delta_ij/3`, `tau_ij = -2 mu_t S_ij`, and shear production `P = -tau_ij S_ij`
* rename "SGS" to "SFS" in code

ewquon committed:
Add additional scalar profile output for completeness

ewquon committed:
Fix sign issue with const g (abs val) vs vector g

ewquon committed:
FIX: Don't recalc eddy viscosity part of KH

ewquon committed:
Update mu_turb at every RK stage, make mixing length calc consistent btwn mu_turb and KH

ewquon committed:
Add limit on mixing length for Deardorff

ewquon committed:
Clarify mixing length

ewquon committed:
Bound KE to ensure it is positive (>= eps)

- AMR-Wind uses a clip value of 1e-8 (in equation_systems/tke/tke_ops.H)
- SOWFA uses a lower bound of "small" (i.e., eps, from SOWFA LESModel property
  kMin; small is defined as floatScalarSmall in OpenFOAM/primitives/Scalar/...)

ewquon committed:
Make bounding on SFS KE consistent across mu_turb and diffusion src updates

ewquon committed:
Update default value of RhoKE_0

If using a SFS model that calculates mu_turb ~ sqrt(k) and the SFS KE is
too small, the MOST boundary condition will run into issues because the
extrapolated boundary value is inversely proportional to the eddy
diffusivity and eddy diffusivity is proportional to mu_turb. The boundary
values blowing up will eventually lead to an erronenous arithmetic error.

For reference, SOWFA users typically define k=0.1; AMR-Wind has a hard-coded
default value of 1.0.

ewquon committed:
Make diffusion src for state consistent w/w/o terrain

ewquon committed:
Use bounded e instead of cell_data(RhoKE) in Deardorff mu_turb calc

ewquon committed:
More Deardorff updates...

- make mixing length in ComputeTurbulentViscosity consistent with DiffusionSrcForState
- calculate local mixing-length-dependent dissipation coefficient Ce
- add option Ce_wall to allow a separate value of Ce at the wall, typically to
  increase dissipation to better represent a "wall effect", (e.g., set to 3.9
  to match Deardorff 1980 and numerous subsequent LES studies)

ewquon committed:
Clip entire RhoKE field; revert to once/timestep mu_turb update

- before: calculate mixing length based on clipped local e, but the e field
  remains untouched -- this could create issues for the TKE transport/diffusion
- confirmed that WRF updates the "exchange coefficients" KM, KH only in the
  first RK stage

ewquon committed:
Flip sign of deviatoric stress tensor tau_ij...

To be consistent with the SGS Reynolds stress definition (e.g., Moeng 1984):
  tau_ij = R_ij - R_kk delta_ij/3
where
  R_ij = (u_i' u_j') + (u_i' (u_j)_r) + ((u_i)_r u_j'),
the RHS includes filtered quantities, (.)_r is the resolved-scale field, and
delta_ij is the kronecker delta.

The RHS diffusion src terms have been updated accordingly

Verified that the following solutions are identical (fcompare gives all 0):
* Deardorff w/ no-slip lower wall
* Deardorff w/ MOST lower BC

ewquon committed:
Update docs to reflect change in sign convention for tau (deviatoric sfs stress tensor)

ewquon committed:
Change default to initialize to constant KE rather than RhoKE

KE_0 param to override default of 0.1 (to match SOWFA, AMR-Wind)
RhoKE_0 param retained for backwards compatibility

ewquon committed:
Add lower limit to the mixing length, following WRF

ewquon committed:
Rename SGS-->SFS

ewquon committed:
Add comments

ewquon committed:
Update TKE RHS only once per timestep...

* use ComputeTurbulentViscosity to update the SFS heat flux and dissipation
  prior to the first RK stage
* use precomputed hfx, diss in DiffusionSrcForState_*
* only compute SmnSmn once (to keep production term frozen)

ewquon committed:
whitespace

ewquon committed:
Merge branch 'development' of github.com:erf-model/ERF into updates_for_ABL_VV

Conflicts:
	Source/Diffusion/ComputeStress_T.cpp
	Source/TimeIntegration/ERF_slow_rhs_post.cpp
	Source/TimeIntegration/ERF_slow_rhs_pre.cpp

ewquon committed:
Fix calls to ComputeSmnSmn

ewquon committed:
Remove debugging print

ewquon committed:
Clarify stability-corrected mixing length calc

ewquon committed:
Remove unused mixing length calc

ewquon committed:
More cleanup

AMLattanzi commented on Source/Diffusion/ComputeStress_T.cpp:
tau_33 at L51 and L276 should have a *= -mu_tot to be consistent.

ewquon replied:
Thanks for catching this!

ewquon committed:
Add missing minus signs

ewquon committed:
Add simple Deardorff test

A linearly changing x-velocity profile (du/dz=const) gives an exact solution
for the SFS KE (when the background density field is uniform). Initializing
to the exact solution should result in no change in e over time.

Notes:
- Theoretically, this test could have been run with slip walls but the current
  BC implementation introduces errors in tau13.
- If rho is not constant, then the vertical velocity field drifts due to non-
  uniform mu_t. Since the boundary velocities are fixed, then this vertical
  motion also causes a change in rho.

ewquon committed:
More cleanup

ewquon committed:
Indexing fix

ewquon committed:
Merge branch 'development' into updates_for_ABL_VV

AMLattanzi reviewed (approved):
These rounds of edits pass all the Biollante reg tests (except ABL-Deardorff since it changes the Deardorff turbulence model) so I'm good with moving it into develop and remaking that benchmark.

AMLattanzi committed:
Merge branch 'development' into updates_for_ABL_VV
